[WIP] alloc_pages

misson20000 opened the pull request:
This PR includes some pretty major refactors and reworks of ace_loader.

First of all, `ace.nro` is now `wk_ace.nro`, since it's specific to webkit. There's another binary now too, `sysmod_ace.nro`. This is pretty much the same thing, but without the webkit-specific cleanup, making it suitable for using under sysmodules.

Next, the whole memory mapping system has been ripped out and replaced with `alloc_pages`. The allocator probably isn't very fast, and isn't very cool and good or anything. This cleans up some of the memory management aspects of ace_loader, making it harder to accidentally overwrite something important with an NRO that we're downloading or something. Something to be mindful of at the moment is that if ace_loader's parent config didn't override the heap, I'm just kinda making an (educated) guess as to how much heap space we're using and that shouldn't get added to the alloc_pages arena and overwriting loader_config's values with it. This is really horrible and we might want to do something better, but we're also always overriding heap on hardware and I just did it so I could test under Mephisto.

I haven't changed HTTP loading of NROs yet to use alloc_pages, which is one of the things that needs to be done before this gets merged.

To finish the job for alloc_pages, we need to add a Homebrew ABI key, add bindings in libtransistor, and implement sbrk as a loop of `map(continuous heap region, alloc_pages(min = 0x1000, max = incr))`.

Both `wk_ace` and `sysmod_ace` are working for me with `test_helloworld.nro`. Not sure if `sysmod_ace` would work under WebKit.

misson20000 committed:
refactor ace_loader into wk_ace and sysmod_ace (breaks wk_ace, and sysmod_ace barely works)

misson20000 committed:
redo ace_loader memory system to use an alloc_pages mechanism instead of mapping and add some files that got ignored in the last commit

misson20000 commented on projects/ace_loader/common_memory.c:
```diff
@@ -0,0 +1,299 @@
+		return true;
+	}
+
+	// TODO: check memory permission/type/state
```
TODO: check memory permission/type/state

misson20000 commented on projects/ace_loader/common_setup.c:
```diff
@@ -0,0 +1,132 @@
+		// do not exit for GUIs
+		if(!nro_loaded_count || nro_unload_fail)
+			return 0;
+			}*/
```
fix this

misson20000 commented on projects/ace_loader/http.c:
```diff
@@ -154,6 +154,7 @@ int http_get_file(const char *path, void *buff, uint64_t maxsize)
 
 	// get an answer
 	ret = parse_header(temp, sizeof(temp), &offs, &got);
+	printf("- parse_header > %d\n", ret);
```
remove this

misson20000 commented on projects/ace_loader/memory.c:
```diff
@@ -1,235 +0,0 @@
-	}
-
-	return 0;
-}
```
for other reviewers: this became `wk_mem_heap_cleanup()`

misson20000 commented on projects/ace_loader/memory.c:
```diff
@@ -1,235 +0,0 @@
-	if(!ptr || !wkBase)
-		return 1;
-	return 0;
-}
```
becomes `wk_mem_install_hook`

misson20000 commented on projects/ace_loader/nro.c:
```diff
@@ -149,35 +152,45 @@ result_t nro_load(void *load_base, int in_size)
 
 	// load NRO
-	r = ro_load_nro(&nro_base, load_base, nro_size, load_base + nro_size, bss_size);
+	r = ro_load_nro(&nro_base, nro_blob, nro_size, nro_bss, bss_size);
```
the `nro_base` and `nro_load_base` terminology is really confusing

misson20000 commented on projects/ace_loader/server.c:
```diff
@@ -80,17 +85,20 @@ int server_init()
 	};
 
 	sck = bsd_socket(2, 1, 6); // AF_INET, SOCK_STREAM, PROTO_TCP
+	printf("- server socket: %d, 0x%x, %d\n", sck, bsd_result, bsd_errno);
```
remove this

misson20000 commented on projects/ace_loader/server.c:
```diff
@@ -432,12 +479,12 @@ void func_exec(char *par)
 		printf("- NRO returned 0x%016lX\n", r);
 	} else
-		printf("- get NRO error %i\n", -ret);
+	printf("- get NRO error %i\n", -ret);*/
```
fix this

misson20000 commented on projects/ace_loader/sysmod_main.c:
```diff
@@ -0,0 +1,64 @@
+		printf("- terminating browser...\n");
+		ret = pm_terminate_process_by_title_id(0x0100000000001011);
+		printf("-   => 0x%x\n", ret);
+		pm_finalize();*/
```
clean this up

misson20000 commented on projects/ace_loader/sysmod_main.c:
```diff
@@ -0,0 +1,64 @@
+			loader_config.heap_overridden = true;
+			loader_config.heap_base = ptr;
+			loader_config.heap_size = (malloc(1) - ptr) + 0x8000;
+		}
```
this hack is terrible

misson20000 commented on projects/ace_loader/sysmod_main.c:
```diff
@@ -0,0 +1,64 @@
+		ret = mem_init();
+		if(ret != 0) {
+			return ret;
+		}
```
get main thread handle?

misson20000 commented:
mentioning #102 and #69

kgsws commented:
I do not see any obvious problems.
Also i suspect HTTP 'module' might become obsolete with SD card access.

misson20000 committed:
move alloc_pages to libtransistor, add support for reprobing address space for available blocks, and reimplement sbrk to map pages obtained from alloc_pages

misson20000 committed:
iron out a few remaining bugs in alloc_pages

misson20000 committed:
add alloc pages test and fix split bug in heap probe

misson20000 committed:
hide alloc_pages debug info behind a flag

misson20000 committed:
Merge branch 'development' into feature-split-aceloader

misson20000 committed:
use alloc_pages in surface and test_audio_output

misson20000 committed:
Merge branch 'development' into feature-split-aceloader

misson20000 committed:
alloc_pages bug squashing:
  - add `ap_lock_region` and corresponding HBABI key to protect loader-managed resources in heap
  - improve block state management
  - initialize `alloc_pages` by probing memory if our heap wasn't overridden
  - fix overlap detection in `ap_insert_and_split_block`
  - fix culling of negative size pages
  - change `ap_dump_info` to be able to use `dbg_printf`, since it doesn't depend on newlib/`malloc`/`sbrk`/`alloc_pages`/itself
  - don't accidentally shrink the heap when we're trying to make it bigger
  - AllocPages key wasn't actually being handled properly
  - add alloc_pages test to Makefile
  - sysmod_ace should call `ap_probe_full_address_space` instead of `ap_init`

misson20000 committed:
sysmod_main can dynamically expand the heap. no need to crank it up all the way, and we might want to leave some memory for other processes

misson20000 committed:
don't coalesce allocated blocks and don't use printf when it might not be safe

misson20000 committed:
set up dbg log for ace_loader too (for alloc_pages debugging)

misson20000 committed:
Merge branch 'development' into HEAD

misson20000 committed:
Merge branch 'development' into feature-split-aceloader

misson20000 committed:
use result codes in alloc_pages

misson20000 committed:
move heap cleanup after finalizer

misson20000 committed:
remove http client from ace_loader

misson20000 committed:
documentation for alloc_pages
